multiset_test.go

```diff
@@ -0,0 +1,306 @@
+	copy(data[:], Secp256k1Order.Bytes())
+	_, err := ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("shouldn't be able to parse a multiset bigger with x bigger than the field size: '%v'", err)
+	}
+	data = [64]byte{}
+	copy(data[32:], Secp256k1Order.Bytes())
+	_, err = ParseMultiSet(data)
+	if err == nil {
+		t.Errorf("shouldn't be able to parse a multiset bigger with y bigger than the field size: '%v'", err)
+	}
+	set := NewMultiset()
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("failed generating random data '%v' '%d' ", err, n)
+	}
+	set.Add(data[:])
+
+}
+
+func TestMultiSet_Reset(t *testing.T) {
+	set := NewMultiset()
+	emptySet := NewMultiset()
+	data := [100]byte{}
+	n, err := rand.Read(data[:])
+	if err != nil || n != len(data) {
+		t.Fatalf("failed generating random data '%v' '%d' ", err, n)
+	}
+	set.Add(data[:])
+	if set.Finalize() == emptySet.Finalize() {
+		t.Errorf("expected set to be empty. found: '%x'", set.Finalize())
+	}
+	set.Reset()
+	if set.Finalize() != emptySet.Finalize() {
+		t.Errorf("expected set to be empty. found: '%x'", set.Finalize())
+	}
+}
+
+func TestMultiSetAddRemove(t *testing.T) {
+	list := [loopsN][100]byte{}
+	set := NewMultiset()
+	set2 := set
+	serializedEmpty := set.Serialize()
+	for i := 0; i < loopsN; i++ {
+		data := [100]byte{}
+		n, err := rand.Read(data[:])
+		if err != nil || n != len(data) {
+			t.Fatalf("Failed generating random data '%d' '%v' ", n, err)
+		}
+		set.Add(data[:])
+		list[i] = data
+	}
+	if set.Finalize() == set2.Finalize() {
+		t.Errorf("sets are the same when they should be different: set '%x'\n", set.Finalize())
+	}
+
+	for i := 0; i < loopsN; i++ {
+		set.Remove(list[i][:])
+	}
+	if set.Finalize() != set2.Finalize() {
+		t.Errorf("sets are different when they should be the same: set1: '%x', set2: '%x'\n", set.Finalize(), set2.Finalize())
+	}
+	if set.Serialize() != serializedEmpty {
+		t.Errorf("serialized sets are different when they should be the same: set1: '%x', set2: '%x'\n", set.Serialize(), serializedEmpty)
+	}
+	parsedSet, err := ParseMultiSet(serializedEmpty)
+	if err != nil {
+		t.Errorf("error: '%v' happened when parsing: '%x'", err, serializedEmpty)
+	} else if parsedSet.Finalize() != set.Finalize() {
+		t.Errorf("sets are different when they should be the same: set1: '%x', parsedSet: '%x'\n", set.Finalize(), parsedSet.Finalize())
+	}
+}
+
+func BenchmarkMultiSet_Add(b *testing.B) {
+	b.ReportAllocs()
+	list := make([][100]byte, b.N)
+	for i := 0; i < b.N; i++ {
+		n, err := rand.Read(list[i][:])
+		if err != nil || n != len(list[i]) {
+			b.Fatalf("Failed generating random data '%d' '%v' ", n, err)
+		}
+	}
+	set := NewMultiset()
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		set.Add(list[i][:])
+		tmpSer := set.Serialize()
+		tmpSet, err := ParseMultiSet(tmpSer)
+		if err != nil || tmpSet.Finalize() != set.Finalize() {
+			panic("bad benchmark")
+		}
+	}
+	if set == NewMultiset() { // To prevent optimizing out the loop
```

Comment:
Is this a real scenario that really happens?

If yes, please note that once you do that NewMultiset returns a pointer, you might had this functionality broken.

Reply:
I honestly don't know :)
I know this is in fact a problem I've encountered many times in C/C++ and Rust, don't know how aggressive is Go's optimizer and if it gives special treatment for benchmarks

Example: https://github.com/bitcoin-core/secp256k1/pull/667 https://github.com/bitcoin-core/secp256k1/pull/678